### PR TITLE
[FLINK-30516] Introduce files table in table store

### PR DESCRIPTION
**Pull request by zjureel**
Introduce files table in Table Store

### CODE REVIEW

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/CoreOptions.java [hunk 1]
```diff
@@ -401,6 +401,12 @@ public class CoreOptions implements Serializable {
                     .withDescription(
                             "Whether to create underlying storage when reading and writing the table.");
 
+    public static final ConfigOption<Long> FILES_SNAPSHOT_ID =
+            ConfigOptions.key("files.snapshot-id")
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/CoreOptions.java:405]: Can we just reuse `scan.snapshot-id`?

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/system/FilesTable.java [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+public class FilesTable implements Table {
+
+    private static final long serialVersionUID = 1L;
+
+    public static final String FILES = "files";
+
+    public static final RowType TABLE_TYPE =
+            new RowType(
+                    Arrays.asList(
+                            new DataField(0, "snapshot_id", new BigIntType(true)),
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/system/FilesTable.java:77]: add partition and bucket?

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/system/FilesTable.java [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+                            new DataField(4, "level", new IntType(false)),
+                            new DataField(5, "record_count", new BigIntType(false)),
+                            new DataField(6, "file_size_in_bytes", new BigIntType(false)),
+                            new DataField(
+                                    7,
+                                    "null_value_counts",
+                                    SerializationUtils.newStringType(false)),
+                            new DataField(
+                                    8,
+                                    "lower_value_bounds",
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/system/FilesTable.java:91]: maybe maxKey and minKey are more important that bounds.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/system/FilesTable.java [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+                                    SerializationUtils.newStringType(false)),
+                            new DataField(
+                                    9,
+                                    "upper_value_bounds",
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/system/FilesTable.java:95]: maybe rename this to something thing like `min max stats`

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/system/FilesTable.java [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+                String fieldName = fieldNames.get(i);
+                FieldStats fieldStats = fieldStatsArray[i];
+                nullValueCounts.put(fieldName, fieldStats.nullCount());
+                lowerValueBounds.put(fieldName, fieldStats.minValue());
+                upperValueBounds.put(fieldName, fieldStats.maxValue());
+            }
+
+            return GenericRow.of(
+                    snapshotId,
+                    BinaryString.fromString(dataFileMeta.fileName()),
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/system/FilesTable.java:284]: schema and data are not consistent.
You should add tests for this.

**Review by JingsongLi — COMMENTED** [review]
Thanks @zjureel for the contribution. Left some comments.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/system/FilesTable.java [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+            TableSchema dataSchema = schemaManager.schema(dataFileMeta.schemaId());
+            RowType keysType = dataSchema.logicalTrimmedPrimaryKeysType();
+            RowDataToObjectArrayConverter keyConverter =
+                    keysType.getFieldCount() > 0
+                            ? new RowDataToObjectArrayConverter(
+                                    dataSchema.logicalTrimmedPrimaryKeysType())
+                            : new RowDataToObjectArrayConverter(dataSchema.logicalRowType());
+
+            int[] indexMapping =
+                    SchemaEvolutionUtil.createIndexMapping(
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/system/FilesTable.java:277]: I don't want to expose `SchemaEvolutionUtil` things here.
Can we extract a some class like `FieldStatsConverters` to do this work?
```
class FieldStatsConverters {
   FieldStatsConverters(Function<Long, TableSchema> schemas, long tableSchemaId) {...}
   FieldStatsArraySerializer get0rCreate(long fileSchemaId);
}
```

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/stats/FieldStatsConverters.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import org.apache.flink.table.store.file.schema.TableSchema;
+import org.apache.flink.table.store.types.DataField;
+import org.apache.flink.table.store.types.RowType;
+
+import java.util.List;
+import java.util.function.Function;
+
+/** Converters to create field stats array serializer. */
+public class FieldStatsConverters {
+    private final Function<Long, TableSchema> schemas;
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/stats/FieldStatsConverters.java:32]: I think this class can maintain a `CocurrentMap` here to cache `FieldStatsArraySerializer`s.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/stats/FieldStatsConverters.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    private final Function<TableSchema, List<DataField>> schemaFields;
+
+    public FieldStatsConverters(Function<Long, TableSchema> schemas) {
+        this(schemas, TableSchema::fields);
+    }
+
+    public FieldStatsConverters(
+            Function<Long, TableSchema> schemas,
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/stats/FieldStatsConverters.java:40]: Maybe a `Function<Long, List<DataField>>`?

**Comment by zjureel** [comment]
Thanks @JingsongLi I have updated the code, please have a look again, THX

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/stats/FieldStatsConverters.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+public class FieldStatsConverters {
+    private final Function<Long, List<DataField>> schemaFields;
+    private final ConcurrentMap<SchemaIdKey, FieldStatsArraySerializer> serializers;
+
+    public FieldStatsConverters(Function<Long, List<DataField>> schemaFields) {
+        this.schemaFields = schemaFields;
+        this.serializers = new ConcurrentHashMap<>();
+    }
+
+    public FieldStatsArraySerializer getOrCreate(long tableSchemaId, long dataSchemaId) {
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/stats/FieldStatsConverters.java:42]: What do you think about my original thought:
```
class FieldStatsConverters {
   FieldStatsConverters(Function<Long, List<DataField>> schemas, long tableSchemaId) {...}
   FieldStatsArraySerializer get0rCreate(long fileSchemaId);
}
```

Maybe we don't need a so flexable `SchemaIdKey`?

**Review by JingsongLi — COMMENTED** [review]
Thanks @zjureel  for the update! I still have some comments on Converters.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/AppendOnlyFileStoreScan.java [hunk 10]
```diff
@@ -102,25 +101,6 @@ protected boolean filterByStats(ManifestEntry entry) {
     /** Note: Keep this thread-safe. */
     private FieldStatsArraySerializer getFieldStatsArraySerializer(long schemaId) {
         return schemaRowStatsConverters.computeIfAbsent(
```
[the remark this answers is not in the document]
Comment (zjureel) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/AppendOnlyFileStoreScan.java:103]: I think it's cool, I indeed found there are two caches in the `StoreScan`

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/stats/FieldStatsConverters.java (hunk 9 above)
Comment (zjureel) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/stats/FieldStatsConverters.java]: LGTM, done

**Comment by zjureel** [comment]
UT fails due to a known issue https://issues.apache.org/jira/browse/FLINK-30750

**Review by JingsongLi — APPROVED** [review]
Looks good to me!

I will merge and rebase later.

**Pull request merged**